### PR TITLE
Dante Editor

### LINKED ISSUE #9: Editor for Articles
We should choose some editor to write articles.
Results of my investigation:

## [Dante2](https://github.com/michelson/dante2)
[my example](https://github.com/babajka/babajka-frontend/blob/dante-editor/components/articles/Editor/)

dante-test

### pros:
* beautiful & simple ui
* meduim like
* good work with images / alignment
* build on [draft.js](https://draftjs.org/), so save content in database as `json`, rather that `html` is more clean way
### cons:
* not universal ([fails when running on server side](https://github.com/michelson/dante2/issues/90))
### solutions
* fork repo & fix all ssr issues

## [Megadraft](https://github.com/globocom/megadraft)
[my example](https://github.com/babajka/babajka-frontend/blob/megadraft-editor/components/articles/Editor.js)

### pros:
* `json` instead `html`
* work with ssr
### cons:
* work with ssr with a lot of warnings
* imho ugly ui
* bad image handling
* no documentation
### solutions:
* customizing it with plugins (it should be easy) and our markup
* not sure about ssr

## [Quill](https://zenoamaro.github.io/react-quill/)
[my example](https://github.com/babajka/babajka-frontend/blob/editor-quill/components/articles/Editor.js)
* you can check that server return rendered content

### pros:
* fully documented
* beautiful & handy ui
### cons:
* state as `html`
* no-ssr, so I render editor only on client (but as we save `html` in db, we get articles seo-optimized)
* conflict with `bulma` styles:
  screen shot 2018-01-03 at 15 42 54
* very bad image support (on first sight)
* same images in database as is in `base64` 😭 
### solutions:
* customizing it with plugins (it should be easy) and our markup

### [alloy-editor](https://github.com/Footage-Firm/alloy-editor)
[my example](https://github.com/babajka/babajka-frontend/blob/alloy-editor/components/articles/Editor.js)

 screen shot 2018-01-03 at 23 49 31

### pros:
* work with ssr withoit big hacks
* it works!

### cons:
* a lot of pain with static files
* not handling content saving
* the only way to save -> manually get `.innerHtml` of node :dizzy_face:
  
There are a lot of other editors that I didn't check yet:
* [Froala](https://github.com/froala/react-froala-wysiwyg)
* [Medium Clones](http://howtox.com/medium-editor-clones-in-js/)
* [More draft.js editors](https://github.com/nikgraf/awesome-draft-js)

What should we choose and do? Help please

### PR DESCRIPTION
**Pull request by Drapegnik**
fix #9 

## Dante2 Editor
screencapture-local-wir-by-3000-be-editortest-2018-06-10-20_57_50

check out it [here](http://local.wir.by:3000/be/editorTest)

## known issues
* styles overriding (Dante break our fonts, images, and with our styles broken Dante tooltips)
screen shot 2018-06-10 at 20 56 00

## done
* [x] fork & fix Dante2
* [x] made it work with ssr
* [x] replace old Quill Editor
* [x] more ui fixes:
screen shot 2018-06-17 at 00 44 12
screen shot 2018-06-17 at 00 44 16
screen shot 2018-06-17 at 00 50 40

## todo
* [x] add image uploading (on cloudinary or just by link) - need rewrite Dante Image Block
* [x] fix styles

## good news
* now article `content` is `draft-js`'s `EditorState` - just `json`
* we can always easy change look for every articles (with [`redraft`](https://github.com/lokiuz/redraft)):
```js
...
  blocks: {
    unstyled: (children) => children.map(child => <p>{child}</p>),
    blockquote: (children) => <blockquote >{addBreaklines(children)}</blockquote>,
    'header-one': (children) => children.map(child => <h1>{child}</h1>),
    'header-two': (children) => children.map(child => <h2>{child}</h2>),
    // You can also access the original keys of the blocks
    'code-block': (children, { keys }) => <pre style={styles.codeBlock} key={keys[0]} >{addBreaklines(children)}</pre>,
    // or depth for nested lists
    'unordered-list-item': (children, { depth, keys }) => <ul key={keys[keys.length - 1]} class={`ul-level-${depth}`}>{children.map(child => <li>{child}</li>)}</ul>,
    'ordered-list-item': (children, { depth, keys }) => <ol key={keys.join('|')} class={`ol-level-${depth}`}>{children.map((child, index)=> <li key={keys[index]}>{child}</li>)}</ol>,
    // If your blocks use meta data it can also be accessed like keys
    atomic: (children, { keys, data }) => children.map((child, i) => <Atomic key={keys[i] {...data[i]} />),
  },
  /**
   * Entities receive children and the entity data
   */
  entities: {
    // key is the entity key value from raw
    LINK: (children, data, { key }) => <Link key={key} to={data.url}>{children}/>,
  },
  /**
...
```
* since I separate public & edit article page for different bundles & on public article just render `content` with `rendraft` we have following bundle sizes:

| bundle           | `Dante2`                        | `Quill`                         |
|------------------|---------------------------------|---------------------------------|
| `editArticle.js` | **1.74 MB**, (174.86 KB gziped) | **1.21 MB** (156.77 KB gziped)  |
| `article.js`     | **97.12 KB**,(9.75 KB gziped)   | **538.88 KB** (75.31 KB gziped) |

### CODE REVIEW

**Comment by Drapegnik** [comment]
[Dante editor](https://trello.com/c/Gc2TAPUm/199-dante-editor)

**Comment by UladBohdan** [comment]
Ok so about the styling issue in navbar.
My quick investigation yesterday showed that the problem is with fonts (not with i.e. `font weight`). Both we and Dante use `Open Sans` font, and we both fetch it from google fonts. We are importing it as Open Sans with two boldnesses (300, 400) and with cyrillic support:
`@import url('https://fonts.googleapis.com/css?family=Open+Sans:300,400,800&subset=cyrillic');`
Dante imports Open Sans with three boldnesses: 300,400,800 (and it seems like cyrillic support is included by default in there).
After we switch the editor on, it imports Open Sans (for the second time for the project) and the font in navbar switches from Open Sans to Open Sans Bold (which is weird).

It's really hard to test it, because making markup-import changes being delivered to Frontend makes us to push new markup to npm registry every time. 😥
And it should be possible to make changes directly into bundle.css files on Frontend although it only breaks styles for me (for some reason, I guess I'm doing something wrong here).

In case we don't have time for proper investigation, suggestions are:
* to replace Open Sans with something else (either in Dante or in our styles for navbar) to avoid conflicts.
* to ignore the issue as non-admin users will never see it and to fix later.

**Comment by Drapegnik** [comment]
> to ignore the issue as non-admin users will never see it and to fix later.

@UladBohdan, Dante styles loads on Public Article page too, and breaks some fonts & some other our styles.

@tyndria, @andemerie can you help us with it?

File: components/common/Editor/index.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  el: 'dante2-editor',
+  read_only: false,
+  spellcheck: false,
+  debug: false,
+  title_placeholder: 'Title',
+  body_placeholder: 'Write your story',
+  widgets,
+  tooltips,
+  default_wrappers: defaultWrappers,
+  continuousBlocks,
```
Comment (tyndria) [inline components/common/Editor/index.js:21]: is it ok, that config has fields in different styles, like `continuousBlocks` and `default_wrappers`?

File: components/common/Editor/tooltips.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      'ordered-list-item',
+      'code-block',
+      'header-one',
+      'header-two',
+      'header-three',
+      'header-four',
+    ],
+    widget_options: {
+      block_types: [
+        // {label: 'p', style: 'unstyled'},
```
Comment (tyndria) [inline components/common/Editor/tooltips.js:26]: are there any reasons to leave this comment?

File: components/common/Editor/widgets.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  type: 'image',
+  block: ImageBlock,
+  editable: true,
+  renderable: true,
+  breakOnContinuous: true,
+  wrapper_class: 'graf graf--figure',
+  selected_class: 'is-selected is-mediaFocused',
+  selectedFn: block => {
+    const { direction } = block.getData().toJS();
+    switch (direction) {
```
Comment (tyndria) [inline components/common/Editor/widgets.js]: I thought you don't like `switch-case` structure :)

Comment (tyndria) [reply components/common/Editor/widgets.js]: (I don't like)

File: components/common/Editor/widgets.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      case 'wide':
+        return 'sectionLayout--fullWidth';
+      case 'fill':
+        return 'graf--layoutFillWidth';
+      default:
+        return null;
+    }
+  },
+  handleEnterWithoutText(ctx, block) {
+    const { editorState } = ctx.state;
```
Comment (tyndria) [inline components/common/Editor/widgets.js]: `handleEnterWithText` function is doing the same?

**Pull request merged**